pykeepass/kdbx_parsing/factorinfo.py

```diff
@@ -0,0 +1,490 @@
+                if wrapping_key is not None:
+                    break
+            if wrapping_key is None:
+                raise CredentialsError("Cannot find a factor to generate validation info")
+            _, self.validation_out = factor.unwrap_key_part(user_supplied_info=user_supplied_info, unwrapping_key=wrapping_key)
+
+        assert self.validation_in is not None and self.validation_out is not None
+
+        key_validation_type = etree.SubElement(group_element, "ValidationType")
+        key_validation_type.text = self.validation_type
+
+        key_validation_in = etree.SubElement(group_element, "ValidationIn")
+        key_validation_in.text = b64encode(self.validation_in)
+
+        key_validation_out = etree.SubElement(group_element, "ValidationOut")
+        key_validation_out.text = b64encode(self.validation_out)
+
+        if self.challenge is not None:
+            challenge = etree.SubElement(group_element, "Challenge")
+            challenge.text = b64encode(self.challenge)
+
+        for factor in self.factors:
+            factor_element = etree.SubElement(group_element, "Factor")
+            factor.encode(factor_element)
+
+    @staticmethod
+    def decode(group_element):
+        key_validation_type = group_element.xpath("ValidationType")[0].text
+        key_validation_in = b64decode(group_element.xpath("ValidationIn")[0].text)
+        key_validation_out = b64decode(group_element.xpath("ValidationOut")[0].text)
+
+        challenge = None
+        challenges = group_element.xpath("Challenge")
+        if len(challenges) > 0:
+            challenge = b64decode(challenges[0].text)
+
+        factors = []
+        for factor in group_element.xpath("Factor"):
+            factors.append(Factor.decode(factor))
+        return FactorGroup(
+            validation_type=key_validation_type,
+            validation_in=key_validation_in,
+            validation_out=key_validation_out,
+            challenge=challenge,
+            factors=factors
+        )
+
+    def unwrap_key_part(self, user_supplied_info):
+        if self.cached_key_part is not None and user_supplied_info == self.cached_key_part[0]:
+            return self.cached_key_part[1]
+
+        fido2_factors = [x for x in self.factors if isinstance(x, FIDO2Factor)]
+        other_factors = [x for x in self.factors if not isinstance(x, FIDO2Factor)]
+
+        for factor in other_factors:
+            # Try non-FIDO factors first
+            try:
+                unwrapped_part, _ = factor.unwrap_key_part(user_supplied_info=user_supplied_info)
+                if unwrapped_part is not None:
+                    self.cached_key_part = (user_supplied_info, unwrapped_part)
+                    return unwrapped_part
+            except CredentialsError as e:
+                log.error("Factor failed: {}".format(e))
+                continue
+
+        next_challenge = random.randbytes(32)
+
+        if len(fido2_factors) > 0:
```

Comment:
Rather important (and very inconvenient to implement...) feature here. We make one FIDO2 call per authenticator per group.

If there are ten Factor entries, each one of which is a FIDO2 authenticator... that means for each authenticator one would succeed and nine would fail.

If there's one authenticator plugged into the system it would not be great for the user to have to type their PIN ten times.

So instead we gather all the credentials in the group and send them in one call. The authenticator will accept the credential it created, and ignore the others.